crates/turbo-tasks-fs/src/lib.rs

```diff
@@ -343,60 +374,99 @@ impl DiskFileSystem {
+                                            batched_invalidate_path_and_children.insert(source.clone());
+                                            if let Some(parent) = source.parent() {
+                                                batched_invalidate_path_dir.insert(PathBuf::from(parent));
+                                            }
+                                            batched_invalidate_path_and_children.insert(destination.clone());
+                                            if let Some(parent) = destination.parent() {
+                                                batched_invalidate_path_dir.insert(PathBuf::from(parent));
+                                            }
+                                            #[cfg(not(any(target_os = "macos", target_os = "windows")))]
+                                            batched_new_paths.insert(destination.clone());
+                                        } else {
+                                            println!("Rename event does not contain source and destination paths {:#?}", paths);
+                                        }
+                                    }
+                                    // [NOTE] Observing `ModifyKind::Metadata(MetadataKind::Any)` is not a mistake.
+                                    // PACK-2437 addresses some cases notify doesn't emit event to trigger hmr,
+                                    // cases like starting from a file have contents -> clear & save.
+                                    // v4 doesn't emit any debounced event for those,
+                                    // and v6 emits an event with `ModifyKind::Metadata(MetadataKind::Any)`
+                                    // it is unclear why these case are not considered as normal modify, anyway we use certain
+                                    // event to workaround.
+                                    EventKind::Any | EventKind::Modify(ModifyKind::Any | ModifyKind::Metadata(MetadataKind::Any)) => {
+                                        batched_invalidate_path.extend(paths.clone());
+                                        batched_invalidate_path_and_children.extend(paths.clone());
+                                        batched_invalidate_path_and_children_dir.extend(paths.clone());
+                                        for parent in paths.iter().filter_map(|path| path.parent()) {
+                                            batched_invalidate_path_dir.insert(PathBuf::from(parent));
+                                        }
+                                    }
+                                    EventKind::Modify(
+                                        ModifyKind::Metadata(..)
+                                        | ModifyKind::Other
+                                        // hope RenameMode::both are suffecient for the rename event, ignore others
+                                        | ModifyKind::Name(..)
```

Comment:
We should handle the other variants too just in case

Reply:
Other then `both` in renamemode for the name event, there are from|to|any|other. 

we can exclude from|to since `both` should be emitted after consolidating those (otherwise it is probably a bug), any should be handled by debouncer if path can be detected (https://github.com/notify-rs/notify/blob/main/notify-debouncer-full/src/lib.rs#L296-L299) and other is unused.

Reply:
TL:DR debouncer-full impl supposed to wrap & handle those cases to not bubble up unexpectedly for us.

Reply:
I would just handle them in any naive way, just to be safe.

Seems like debouncer can emit a To event if it can match From and To.

Reply:
Yes it emit to, which will be dupe event to both. Do we want multiple invalidation with same change?

Reply:
> Do we want multiple invalidation with same change?

It doesn't matter.

> which will be dupe event to both

It might emit `To` when it can't match it with a `From` event. Not sure when this can happen, but I would feel better if we handle it